### PR TITLE
jitson parser

### PR DESCRIPTION
**Pull request by dnlup**
Write basic plugin implementation.
Add Test, Docs, Benchmark.

Note:
I added `package.json` with an empty author field. I thought it would have been better if you would consider putting this with the other core plugins.

### CODE REVIEW

File: plugin.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+const fp = require('fastify-plugin')
+const jitson = require('jitson')
+
+function plugin (fastify, options, next) {
+  fastify.addContentTypeParser('application/json', { parseAs: 'string' }, function (req, body, done) {
+    const parse = jitson(options)
```
Comment (Eomm) [inline plugin.js]: This line should go outside the `addContentTypeParser` function? In this way, we cache the `parse` object and we can avoid to create it for each json request.
Moreover, if we recreate it each request we are not losing the schema sampling?

I think this part could be optimized with a Map() where the key is the route path and the value the jitson object

WDYT?

**Review by Eomm — COMMENTED** [review]
I would also add to remove the `package-lock.js` file: in a module is not good to add it since the published module on npm will never has it (this file is ignored) and if we have it in the repo, CI and our local env could not reproduce some trouble

File: plugin.js (hunk 1 above)
Comment (mcollina) [reply plugin.js]: Yes, good spot!

Comment (dnlup) [reply plugin.js]: @Eomm You are absolutely right

**Comment by dnlup** [comment]
Refactored.
I am sure it is my fault but I see a big performance degradation.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  "directories": {
+    "test": "test"
+  },
+  "scripts": {
+    "test": "tap test/*.js",
+    "lint": "standard --verbose | snazzy",
+    "bench": "branchcmp -r 2 -g -s \"npm run benchmark\"",
+    "benchmark": "npx concurrently -k -s first \"node ./example/jitson.js\" \"npm run benchmark:request\"",
+    "benchmark:core": "npx concurrently -k -s first \"node ./example/core.js\" \"npm run benchmark:request\"",
+    "benchmark:request": "npx autocannon -c 100 -d 5 -p 10 -m POST -b '{ \"hello\": \"world\" }' -H Content-Type=application/json localhost:3000/"
```
Comment (mcollina) [inline package.json:15]: You'll likely need to do two or three runs to get stable numbers. What's the difference between core and jitson?

Comment (dnlup) [reply package.json:15]: Screenshots of 4th run

```bash
npm run benchmark
```
Screenshot from 2019-03-26 14-16-24

```bash
npm run benchmark:core
```
Screenshot from 2019-03-26 14-17-18

**Comment by mcollina** [comment]
I'm afraid this will never give us the speed we need. It's not really a good idea to use jitson.

However, we can try using https://github.com/mafintosh/turbo-json-parse instead. The overall idea is that, instead of calling `jitson`, we can compile the parser from the JSON Schema of the body itself. Would you like to try that instead?

**Comment by dnlup** [comment]
@mcollina yes! I'll get back to you as soon as I am done

**Comment by dnlup** [comment]
@mcollina Ok, in this [PR](https://github.com/dnlup/fastify-tjp/pull/1) you can find a basic implementation. I haven't added docs and tests because I still see that the performance is not better that the default parser (of course it could be my implementation). If you have time to try it the benchmark scripts are the same as the ones in this repo, otherwise I will attach screenshots.
Closing this in favour of https://github.com/dnlup/fastify-tjp/pull/1.

**Pull request closed without merging**